### ISSUE TITLE
Is it possible to receive dispatch commands from Voice? I am migrating from 2.12.1 to 2.16.1

### ISSUE THREAD

**Issue by alveru** [google/ExoPlayer#10056]
Hi, I am migrating from library 2.12.1 to 2.16.1, I found out the ControlDispatcher was completely removed.

I was using the ControlDispatcher to intercept the voice commands for play/pause to send the appropriate Analytics tracking message. Now I am kind of lost how to detect and intercept the voice commands to track them, in fact, complete media session
commands stopped working because there is not control dispatcher, so i am wondering if there is a tutorial or documentation about how to do this, the google documentation page still makes reference to old control dispatchers.

I read in the ReleaseNotes.md file that i could use a forwarding player, but i am not sure exactly how to do it, this is the old code i used to track the voice commands:

```
class CustomControlDispatcher(val videoPresenter: VideoPresenter) : ControlDispatcher {

    val titleId: String?
        get() =videoPresenter.currentTitle()

    override fun dispatchSetPlayWhenReady(player: Player, playWhenReady: Boolean): Boolean {
        // change state of player manually
        player.playWhenReady = playWhenReady
       // Tracking external commands such as voice commands in AndroidTV
       AnalyticsTracker.track("voice_cmd_play", titleId)
        return true
    }

    override fun dispatchSeekTo(player: Player, windowIndex: Int, positionMs: Long): Boolean {
        //  set position of player manually
        player.seekTo(windowIndex, positionMs)
       // Track in analytics the voice command
       AnalyticsTracker.track("voice_cmd_seek_to", titleId, windowIndex, positionMs)
        return true
    }

```
I just attached this custom control dispatcher to the media session and i started receiving commands, i tested them with:
"adb shell media dispatch pause" and now nothing works, the player doesn't receive the commands. I extended a class from ForwardingPlayer and set its player to my video player but i do not receive any message at all, can somebody give me some light about how to replace the control dispatcher?

This is the class i created:

```
class VoiceCommandsDispatcher(player: Player) : ForwardingPlayer(player) {

    override fun setPlayWhenReady(playWhenReady: Boolean) {
        super.setPlayWhenReady(playWhenReady)

        Log.e("--->", " ---> VoiceCommandsDispatcher: setPlayWhenReady: $playWhenReady")
        }
    }

    override fun seekTo(mediaItemIndex: Int, positionMs: Long) {
        super.seekTo(mediaItemIndex, positionMs)

        Log.e("--->", " ---> VoiceCommandsDispatcher: seekTo: $mediaItemIndex, $positionMs")
    }
}
```
I will really appreciate any help on this

**Comment by marcbaechinger** [comment]
Just to make sure and because you don't tell what you do with your `VoiceCommandsDispatcher`. This should work 

```
mediaSessionConnector.setPlayer(new VoiceCommandsDispatcher(player));
```

When you are saying that you don't receive commands at all then I would put some log statements/break points in the `MediaSessionConnector` itself. Do this in the[ Callback at the end of the class](https://github.com/google/ExoPlayer/blob/release-v2/extensions/mediasession/src/main/java/com/google/android/exoplayer2/ext/mediasession/MediaSessionConnector.java#L1188).

You should see commands arriving there which then are delegated to your forwarding player. If this is not the case, then something goes probably wrong with that adb command `db shell media dispatch pause`. 

You are not specifiying your app with that command, so in some ways the command needs to figure out to  which app/media session to send that command. I would expect this to boil down to a media button event like sent from a BT headset. Alternatively for testing, you could choose to open Assistant on your phone and type or say 'pause' instead. Just for checking that command is not the problem.

As a first step I would start the app before you send the adb command and then I'd verify whether the media session of your app is there as expected:

```
adb shell dumpsys media_session
```

This command list the media sessions the system currently knows. Check whether you find your app's session at the top and that your session is active. 

If this is the case and we still have no luck, then I don't know. :( Probably best that you would create a bug report and upload here so that I can investigate on my end.

**Comment by alveru** [comment]
hi @marcbaechinger , thanks for answering my question.

I was digesting the response, it clarified many things, so, what i did is i removed the ForwardingPlayer class and instead extended from MediaSessionCompat.Callback() and override onPause, onPlay and onSeek as the link you posted, and just added it as the callback, and now it seems to be working !!! 👍 🥇 

the only difference that I see is that 
the onSeekTo function is missing a parameter called "mediaItemIndex", not sure if that was intended in the upgrade? so for now, i left it with the position only, because i was tracking that value too 😖 

Adding the code in case anyone else is asking themselves the same question:

```
mediaSessionConnector?.setPlayer(exoPlayer)
// Detect voice commands
mediaSessionCompat.setCallback(VoiceCommandsDispatcher(player))
                
class VoiceCommandsDispatcher(val player: Player
) : MediaSessionCompat.Callback() {

    override fun onPlay() {
        super.onPlay()
        if (BuildConfig.DEBUG) Log.e("VoiceCommandsDispatcher", " Media Session Dispatcher: onPlay")
    }

    override fun onPause() {
        super.onPause()
        if (BuildConfig.DEBUG) Log.e("VoiceCommandsDispatcher", " Media Session Dispatcher: onPause")
    }

      override fun onSeekTo(pos: Long) {
        super.onSeekTo(pos)
        if (BuildConfig.DEBUG) Log.e("VoiceCommandsDispatcher", " Media Session Dispatcher: onSeekTo: [$pos]")
      }
}
```

**Comment by marcbaechinger** [comment]
I don't think you should do your own `Callback`. With this you need to dispatch all the events yourself. If your `Callback` is receiving the calls, then the `MediaSessionConnector.Callback` should receive the same calls. There must be another issue with the integration that prevents it from working. Doing  your own callback is not a solution that is helpful for other users.

> the onSeekTo function is missing a parameter called "mediaItemIndex", not sure if that was intended in the upgrade?

Look like you mix up the `MediaSession.Callback` with `Player.seekTo`? I don't think there ever was a `mediaItemIndex` parameter in the `MediaSessionCompat.Callback`.

**Comment by alveru** [comment]
Thanks for your reply.

1) so the callback  in the MediaSessionCompat should not be overriden?;  ok, so i removed it from the code to leave it as it was, and then I put a break in the function inside MediaSessionConnector.ComponentListener in the onPause function:

@Override
    public void onPause() {
     if (canDispatchPlaybackAction(PlaybackStateCompat.ACTION_PAUSE)) {    <---- Breakpoint here
        player.pause();
      }
    }


and, execution is stopping correctly at that point with the "adb shell media dispatch pause" command, and is pausing the player now. But, then, i am not sure where exactly should I intercept this message so i can send the analytics to track that user set a voice command to pause/play/seek , since this function (MediaSessionConnector->onPause) is inside the library jar file (2.16.1-sources.jar:com.google.android.exoplayer2.ext.mediasession) Should i extend a class from MediaSessionConnector to override those functions (onPause/onPlay/onSeek)?


2) Regarding the <mediaItemIndex>, i was using the function inside the interface ControlDispatcher

override fun dispatchSeekTo(player: Player, windowIndex: Int, positionMs: Long): Boolean {
}

ControlDispatcher was removed in 2.16.0, and i was using a custom class that implemented the  ControlDispatcher interface to send the analytics tracking there, and i am not sure where to put these trackers  in the new 2.16.1 version.

**Comment by alveru** [comment]
I reviewed again your second response, and changed the approach to use the ForwardingPlayer again, i extended the class, and I am passing this extended class to the MediaSessionConnector to receive the voice commands, and I am using the original ExoPlayer as parameter to the ForwardingPlayer, and i think is working 👍 , the override functions from the media session receive the actions and i can track them in analytics before dispatching them to the exoplayer, this is the code i used:

class VoiceCommandDispatcher(
    val player: Player
) : ForwardingPlayer(player) {

    companion object {
        private val TAG = "VoiceCommandsDispatcher"
    }

    override fun play() {
        if (BuildConfig.DEBUG) Log.e(TAG, " ---> Media Session Dispatcher: onPlay")
        // Do something with play voice command or just call super.play() to forward to player implementation
    }

    override fun pause() {
        if (BuildConfig.DEBUG) Log.e(TAG, " ---> Media Session Dispatcher: onPause")
        // Do something with pause voice command or just call super.pause() to forward to player implementation
    }

    override fun seekTo(mediaItemIndex: Int, positionMs: Long) {
        if (BuildConfig.DEBUG) Log.e(TAG, " ---> Media Session Dispatcher: onSeekTo: [$positionMs]")
        // Do something with the seekTo voice command, or just call super.seekTo(mediaItemIndex, positionMs) to forward to player implementation
    }
}

// Connect command dispatcher to the media session, and the exoplayer to the command dispatcher
val commandDispatcher = VoiceCommandDispatcher(exoplayer)  <--- Dispatcher will forward media session actions to this exoplayer
val mediaSessionConnector = MediaSessionConnector(mediaSessionCompat)
mediaSessionConnector.setPlayer(commandDispatcher)        <--- This cmd dispatcher will intercept media session actions

I think it has the correct approach as you described originally. Thanks for the support.
